ttnn/cpp/ttnn/operations/ccl/ccl_op_fusion.cpp

```diff
@@ -0,0 +1,179 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "tt_metal/host_api.hpp"
+#include "tt_metal/impl/program/program.hpp"
+#include "ttnn/operations/ccl/ccl_op_fusion.hpp"
+
+namespace ttnn {
+namespace experimental {
+namespace ccl {
+
+void AllGatherFusedOpSignaler::init_fused_op(
+    const std::vector<CoreCoord>& fused_op_receiver_cores_noc,
+    const std::vector<uint32_t>& fused_op_receiver_signal_semaphores
+) {
+    this->fused_op_receiver_cores_noc = fused_op_receiver_cores_noc;
+    this->fused_op_receiver_signal_semaphores = fused_op_receiver_signal_semaphores;
+    this->num_fused_op_cores_to_signal = fused_op_receiver_cores_noc.size();
+
+    initialized_fused_op = true;
+}
+
+void AllGatherFusedOpSignaler::init_all_gather(
+    Program& program,
+    Device const* device,
+
+    CoreRangeSet const& all_gather_workers,
+    std::vector<CoreCoord>& all_gather_worker_cores
+) {
+    // Create the sync semaphore for the all gather workers
+    this->all_gather_worker_sync_semaphore = CreateSemaphore(program, all_gather_workers, 0);
+
+    // Get the noc coords for the all gather workers
+    this->all_gather_worker_cores_noc.clear();
+    for (const auto& core : all_gather_worker_cores) {
+        this->all_gather_worker_cores_noc.push_back(device->worker_core_from_logical_core(core));
+    }
+    initialized_all_gather = true;
+}
+
+void AllGatherFusedOpSignaler::emit_all_gather_fused_op_rt_args(
+    std::vector<uint32_t>& rt_args,
```

Comment:
could you rename it at least to be out_rt_args?
Or better just return std::vector<uint32_t>.
Overall passing out parameter to be first is a little bit missleading.